### PR TITLE
Move `test::Bencher` to a new (unstable) `std::bench` module

### PR DESCRIPTION
**Pull request by SimonSapin**
This is a first step toward stabilizing it and the `#[bench]` attribute: #66287

To avoid also moving much of the `test` crate `Bencher` is now only responsible for runnning user code and measuring the time it takes, not anymore for doing statistical analysis. This separation is based on `&mut dyn FnMut` callbacks.

This introduces dynamic dispatch, which in general could affect performance characteristics of a program. However I expect benchmarking results not to be affected here since the {`Instant::new`; user code; `Instant::elapsed`} sequence is kept monomorphized and not crossing a dynamic dispatch boundary.

This also adds a lifetime parameter to the `Bencher` struct, which is is technically a breaking change to an unstable type. I expect the impact to be low on existing users: only those with `#[deny(warnings)]` or `#[deny(elided_lifetimes_in_paths)]` would need to change their code. (See second commit.) This lint is `allow` by default.

### CODE REVIEW

File: src/libstd/lib.rs [hunk 1]
```diff
@@ -454,6 +454,7 @@ pub mod f64;
 pub mod thread;
 pub mod ascii;
 pub mod backtrace;
+pub mod bench;
```
Comment (Centril) [inline src/libstd/lib.rs]: Shouldn't the module also have a stability annotation?

Comment (SimonSapin) [reply src/libstd/lib.rs]: Yes? It’s near the top of `libstd/bench.rs`, like for most other modules under `std::`

Comment (Centril) [reply src/libstd/lib.rs]: I saw that, but I guess it isn't necessary on the mod item itself?

Comment (SimonSapin) [reply src/libstd/lib.rs]: This (inner) attribute is already on the mod item.  If we switch to inline module syntax, this PR is equivalent to:

```rust
pub mod bench {
    #![unstable(…)]
    // …
}
```

Which is equivalent to:


```rust
#[unstable(…)]
pub mod bench {
    // …
}
```

**Comment by elichai** [comment]
May I ask why dynamic dispatch and not generic?

**Comment by SimonSapin** [comment]
Dynamic dispatch avoids forcing a type parameter on all functions with the `#[bench]` attribute. And passing `impl FnMut` to another `impl FnMut` would require the outer one to be a higher-kinded type, which don’t exist in the language (at least today).

**Comment by JohnCSimon** [comment]
Ping from triage:
@rkruppe Can you please review this PR? thank you
cc: @SimonSapin @elichai @Centril

**Comment by SimonSapin** [comment]
This PR has served its initial purpose of demonstrating feasibility. I’ll close it now to avoid keeping it open for a long time on triage’s radar. We can re-open it if and when we decide in https://github.com/rust-lang/rust/issues/66287 or in an RFC that this the approach we want.

**Pull request closed without merging**